### ISSUE TITLE
Email and push notifications randomly stop sending - requires server restart each time to fix

### ISSUE THREAD

**Issue by dkoo761** [RocketChat/Rocket.Chat#21039]
### Description:

Email and push notifications randomly stop sending for no apparent reason. The notifications are inserted into the rocketchat `email_notifications_queue` collection in the DB but are never processed. No errors in the logs. Once these notifications stop sending, the only way to get them to send again is to restart the server. Upon restart, everything that was sitting in email_notifications_queue is then sent all at once.

After looking at the code and adding some logging into the NotificationQueue.ts file, it appears that the `worker()` function randomly stops running. The server then needs to be restarted so that `Notification.initWorker()` runs again to kick off the `worker()` job. In my case, this happens anywhere from a few hours to a few days but at least happens once a week and on a very low volume instance of RocketChat.

I think the implementation of the `worker()` logic is quite fragile as it has no way to recover if something goes wrong. Perhaps it would make sense to use the `littledata:synced-cron` package for this since it's basically a scheduled job. Or at least use the synced-cron package to check say every 5 minutes to make sure the worker is still running and if not, call `Notification.initWorker()` to start it up again.

### Steps to reproduce:

1. Send a DM to an offline user.
2. See if the offline user got the email and push notification.

### Expected behavior:

Offline user got the notification(s)

### Actual behavior:

Sometimes they get the notification(s), sometimes they don't

### Server Setup Information:

- Version of Rocket.Chat Server: 3.7.4
- Operating System: Ubuntu 18.04
- Deployment Method: Manual (https://docs.rocket.chat/installation/manual-installation/ubuntu)
- Number of Running Instances: 1
- DB Replicaset Oplog: Enabled
- NodeJS Version: 12.14.0 - x64
- MongoDB Version: 4.2.11

### Client Setup Information

- Desktop App or Browser Version: Chrome 87.0.4280.88 (Official Build) (x86_64)
- Operating System: Mac OS 10.15.7 Catalina

**Comment by pierreozoux** [comment]
I have the same behavior.

I though that https://github.com/RocketChat/Rocket.Chat/pull/20605 would fix it, but still the same. I have 3.12.1, I have log level info, and I see nothing.

If there is anything I can do help investigate, please let me know!

**Comment by pierreozoux** [comment]
Hum.. I was happy to think it was related to https://forums.rocket.chat/t/messages-not-displaying-and-greyed-out/2035/17 and that this person probably has the fix for us :) But you are the same person..

I'm applying your fix, I'll see, but then my hope is lower than previously :)

**Comment by dkoo761** [comment]
@pierreozoux Yes that was me too....so sorry to disappoint you! 😂 I was hoping that my solution from that forum post would work for offline email notifications but unfortunately they were still happening.

In NotificationQueue.ts I see that there are some calls to `this.worker(...)` that are missing `await`. Since `this.worker()` is async, that might be related but I'm not sure. The code is using various async functions (some with `await`, some without) + recursion + setTimeout so I'm guessing there's some edge case in there somewhere that isn't being handled.

As I mentioned, I think the safest approach is to have a separate job that somehow checks to make sure that the worker is still running. For example, the worker could save a lastRunAt timestamp somewhere and the job could check that to make sure it's run in the last N minutes.

**Comment by pierreozoux** [comment]
You are right, a job is a job and should be a separate process. I guess their new enterprise offering with microservices has this. In this sense, rails is quiet nicely architected. But again, we could search how is the meteor community doing to run fibers separated from the main webserver ( a quick research didn't bring quick gold :) ).

What is complicated for me is that this side process doesn't send any log to say out loud that it failed.

**Comment by dkoo761** [comment]
@pierreozoux Just to clarify, the `littledata:synced-cron` (https://github.com/percolatestudio/meteor-synced-cron) is a common way to handle scheduled jobs in Meteor apps. It works with multiple instances of your app so that even if you have N instances, the job only runs once at the scheduled time or interval, not N times.

**Comment by cprice-kgi** [comment]
We are experiencing this exact same issue. Has anyone made progress in resolving it?

**Comment by dkoo761** [comment]
My solution was to modify the RocketChat source code in `app/lib/server/lib/sendNotificationsOnMessage.js` to send the email immediately rather than queueing it. This has worked perfectly ever since. However I'm running on an old version 3.7.1, so I'm not sure if it's still relevant to the latest version of RC.
